[Quirk] maps.google.com: [iPad] unable to scroll on google maps print page

jesxilin opened the pull request:
#### 5988703a14ecb92b02c407f7a319ac19a88f41c3
<pre>
[Quirk] maps.google.com: [iPad] unable to scroll on google maps print page
<a href="https://bugs.webkit.org/show_bug.cgi?id=282721">https://bugs.webkit.org/show_bug.cgi?id=282721</a>
<a href="https://rdar.apple.com/67358928">rdar://67358928</a>

Reviewed by Wenson Hsieh.

Currently, we cannot scroll on maps.google.com via touch b/c there is a `touch-action: none`
on `&lt;body&gt;`. In their code, they recieve all the expected touch events but they do not
move the page to &quot;scroll&quot;.

There is a `div .UJrD4b` which touch events are happening on.
There is another `div #print` which is the grandparent of `div .UJrD4b`.
`div #print` has `display: none` on all pages except the printing pages.
Based on the pointer events “Determining supported direct manipulation behavior”spec,
we travel up to the inclusive ancestor which is `&lt;body&gt;`
that has a `touch-action: none` style.

1. To quirk and make touches affect scrolling, we set `&lt;body&gt;`
to have `touch-action: auto` instead. To do so, we check if `.PUtLdf` is present on body.
While this seems to be a generated class name, the proper quirk for this would require
breaking CSS inheritance since `&lt;body&gt;` would now depend on the `display` styling of the
descendant `div #print`, resulting in a circular dependency.
2. Remove `#if ENABLE(TOUCH_EVENTS)` for `Quirks::isGoogleMaps()`.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGoogleMapsScrollingQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: <a href="https://commits.webkit.org/286455@main">https://commits.webkit.org/286455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/979ef653b5922daf9870f9a0d5ef50b81af2829a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80532 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27286 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78136 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3345 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59612 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17761 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49496 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65286 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39970 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46895 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22770 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81981 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3389 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2170 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67837 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3543 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67146 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11097 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9217 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3339 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6145 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3360 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4298 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -968,6 +968,14 @@ void Adjuster::adjustForSiteSpecificQuirks(RenderStyle& style) const
 
+#if PLATFORM(IOS_FAMILY)
+    if (m_document->quirks().needsGoogleMapsScrollingQuirk()) {
+        static MainThreadNeverDestroyed<const AtomString> className("PUtLdf"_s);
```
Do we know how stable this is? It seems like the kind of thing that maybe even just re-deploying the software might change?

jesxilin replied:
Yeah... If I understand correctly, it would be better to check for `div .UJrD4b` which should always be there on the page and what I initially set to do, but I was having trouble accessing the `<body>` that we needed to change the style on, which is why I choose to search for this class name since it's particular to the print page.

whsieh replied:
From inspecting Google Maps' DOM, I wonder if it's more stable to check for the existence of a `div` with `id=print`? That would avoid using any element IDs that (at least) _seem_ like they might be autogenerated or rotate regularly.

hortont424 commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -968,6 +968,14 @@ void Adjuster::adjustForSiteSpecificQuirks(RenderStyle& style) const
+#if PLATFORM(IOS_FAMILY)
+    if (m_document->quirks().needsGoogleMapsScrollingQuirk()) {
+        static MainThreadNeverDestroyed<const AtomString> className("PUtLdf"_s);
+        if (m_element == m_document->body() && const_cast<Element*>(m_element.get())->classList().contains(className))
```
I think you can avoid this const_cast by using Element's `hasClassName` method? (just make sure you grab ElementInlines.h)

hortont424 commented on Source/WebCore/page/Quirks.h:
```diff
@@ -249,6 +250,7 @@ class Quirks {
     mutable std::optional<bool> m_needsFullscreenObjectFitQuirk;
     mutable std::optional<bool> m_shouldAvoidPastingImagesAsWebContent;
     mutable std::optional<bool> m_mayNeedToIgnoreContentObservation;
+    mutable std::optional<bool> m_needsGoogleMapsScrollQuirk;
```
"scroll" or "scrolling"? not sure why the inconsistency

jesxilin replied:
Will fix the inconsistency!

pxlcoder replied:
I think we should use "scrolling"? Like `needsWeChatScrollingQuirk`.

pxlcoder commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -968,6 +968,17 @@ void Adjuster::adjustForSiteSpecificQuirks(RenderStyle& style) const
+#if PLATFORM(IOS_FAMILY)
+    if (m_document->quirks().needsGoogleMapsScrollQuirk()) {
+        static MainThreadNeverDestroyed<const AtomString> idValue("print"_s);
+        if (m_element == m_document->body()) {
```
```suggestion
        if (is<HTMLBodyElement>(*m_element)) {
```

jesxilin committed:
[Quirk] maps.google.com: [iPad] unable to scroll on google maps print page
https://bugs.webkit.org/show_bug.cgi?id=282721
rdar://67358928

Reviewed by Wenson Hsieh.

Currently, we cannot scroll on maps.google.com via touch b/c there is a `touch-action: none`
on `<body>`. In their code, they recieve all the expected touch events but they do not
move the page to "scroll".

There is a `div .UJrD4b` which touch events are happening on.
There is another `div #print` which is the grandparent of `div .UJrD4b`.
`div #print` has `display: none` on all pages except the printing pages.
Based on the pointer events “Determining supported direct manipulation behavior”spec,
we travel up to the inclusive ancestor which is `<body>`
that has a `touch-action: none` style.

1. To quirk and make touches affect scrolling, we set `<body>`
to have `touch-action: auto` instead. To do so, we check if `.PUtLdf` is present on body.
While this seems to be a generated class name, the proper quirk for this would require
breaking CSS inheritance since `<body>` would now depend on the `display` styling of the
descendant `div #print`, resulting in a circular dependency.
2. Remove `#if ENABLE(TOUCH_EVENTS)` for `Quirks::isGoogleMaps()`.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGoogleMapsScrollingQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: https://commits.webkit.org/286455@main
